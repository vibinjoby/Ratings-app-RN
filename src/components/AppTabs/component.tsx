import React from 'react'
import { View, Text, TouchableOpacity, ButtonProps, ViewProps } from 'react-native'

import styles from './styles'

export interface AppTabsProps {
  tabs: string[]
  selectedTab: number
  onPress: (tab: number) => void
}

interface TabProps {
  tabsLength: number
  tabIndex: number
  selectedTab: number
  onPress: ButtonProps['onPress']
  tabText: string
}

const stylesMap: Record<number, ViewProps['style']> = {
  0: styles.leftTab,
  1: styles.middleTab,
  2: styles.rightTab,
}

const testIdMap: Record<number, string> = {
  0: 'left',
  1: 'middle',
  2: 'right',
}

/* If the tab size is 2 and when the tab index is 1 increment it to 2, so that middle tab becomes right tab */
const Tab = ({ tabsLength, tabIndex, selectedTab, onPress, tabText }: TabProps) => {
  const tabMapIndex = tabsLength === 2 && tabIndex === 1 ? tabIndex + 1 : tabIndex
  return (
    <TouchableOpacity testID={`${testIdMap[tabMapIndex]}Tab`} onPress={onPress} activeOpacity={0.7}>
      <View
        style={[
          stylesMap[tabMapIndex],
          styles.tabs,
          selectedTab === tabIndex && styles.selectedTab,
        ]}
      >
        <Text
          testID={`tab${tabIndex + 1}Txt`}
          style={[styles.tabTxt, selectedTab === tabIndex && styles.selectedTabText]}
        >
          {tabText}
        </Text>
      </View>
    </TouchableOpacity>
  )
}

const AppTabs: React.FC<AppTabsProps> = ({ tabs, selectedTab, onPress }: AppTabsProps) => (
  <View style={styles.container}>
    {tabs.map((tab, index) => (
      <Tab
        tabsLength={tabs.length}
        tabIndex={index}
        selectedTab={selectedTab}
        key={index}
        tabText={tab}
        onPress={() => onPress(index)}
      />
    ))}
  </View>
)

export default AppTabs
