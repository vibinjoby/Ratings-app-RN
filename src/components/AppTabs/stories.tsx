import React, { useState } from 'react'
import { StyleSheet, View } from 'react-native'
import { storiesOf } from '@storybook/react-native'

import AppTabs from './component'

const Component = () => {
  const [selectedTab, setSelectedTab] = useState(0)
  return (
    <>
      <View style={styles.container}>
        <AppTabs
          selectedTab={selectedTab}
          tabs={['Customer', 'Owner', 'Admin']}
          onPress={(tab) => setSelectedTab(tab)}
        />
      </View>
      <View style={styles.container}>
        <AppTabs
          selectedTab={selectedTab}
          tabs={['Customer', 'Owner']}
          onPress={(tab) => setSelectedTab(tab)}
        />
      </View>
    </>
  )
}

storiesOf('Components', module).add('AppTabs', () => <Component />)

const styles = StyleSheet.create({
  container: {
    alignSelf: 'center',
    marginTop: 20,
  },
})
