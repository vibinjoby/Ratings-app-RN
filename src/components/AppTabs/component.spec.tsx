import React from 'react'
import { fireEvent, render } from '@testing-library/react-native'

import AppTabs from './component'

describe('AppTabs', () => {
  const mockFn = jest.fn()
  const SELECTED_TAB = 0
  const TAB1TEXT = 'Customer'
  const TAB2TEXT = 'Owner'
  const TAB3TEXT = 'Admin'

  it('should render', () => {
    const { toJSON } = render(
      <AppTabs selectedTab={SELECTED_TAB} tabs={[TAB1TEXT, TAB2TEXT, TAB3TEXT]} onPress={mockFn} />,
    )
    expect(toJSON()).toMatchSnapshot()
  })

  it('should work for three tabs', () => {
    const { getByTestId } = render(
      <AppTabs selectedTab={SELECTED_TAB} tabs={[TAB1TEXT, TAB2TEXT, TAB3TEXT]} onPress={mockFn} />,
    )

    expect(getByTestId('tab1Txt').props.children).toBe(TAB1TEXT)
    expect(getByTestId('tab2Txt').props.children).toBe(TAB2TEXT)
    expect(getByTestId('tab3Txt').props.children).toBe(TAB3TEXT)

    fireEvent.press(getByTestId('leftTab'))
    fireEvent.press(getByTestId('middleTab'))
    fireEvent.press(getByTestId('rightTab'))
  })

  it('should work for two tabs', () => {
    const { getByTestId } = render(
      <AppTabs selectedTab={SELECTED_TAB} tabs={[TAB1TEXT, TAB2TEXT]} onPress={mockFn} />,
    )

    expect(getByTestId('tab1Txt').props.children).toBe(TAB1TEXT)
    expect(getByTestId('tab2Txt').props.children).toBe(TAB2TEXT)

    fireEvent.press(getByTestId('leftTab'))
    fireEvent.press(getByTestId('rightTab'))
  })
})
