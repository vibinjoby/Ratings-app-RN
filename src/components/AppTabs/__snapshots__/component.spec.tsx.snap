// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppTabs should render 1`] = `
<View
  style={
    Object {
      "alignSelf": "flex-start",
      "flexDirection": "row",
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="leftTab"
  >
    <View
      style={
        Array [
          Object {
            "borderBottomLeftRadius": 40,
            "borderTopLeftRadius": 40,
            "marginRight": 2,
          },
          Object {
            "alignItems": "center",
            "backgroundColor": "#E2E2E2",
            "borderColor": "#fff",
            "borderWidth": 1,
            "elevation": 3,
            "paddingHorizontal": 14,
            "paddingVertical": 8,
            "shadowColor": "#181725",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.22,
            "shadowRadius": 2.22,
          },
          Object {
            "backgroundColor": "#FF5823",
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Poppins-Regular",
              "fontSize": 16,
            },
            Object {
              "color": "#fff",
            },
          ]
        }
        testID="tab1Txt"
      >
        Customer
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="middleTab"
  >
    <View
      style={
        Array [
          Object {
            "marginRight": 2,
          },
          Object {
            "alignItems": "center",
            "backgroundColor": "#E2E2E2",
            "borderColor": "#fff",
            "borderWidth": 1,
            "elevation": 3,
            "paddingHorizontal": 14,
            "paddingVertical": 8,
            "shadowColor": "#181725",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.22,
            "shadowRadius": 2.22,
          },
          false,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Poppins-Regular",
              "fontSize": 16,
            },
            false,
          ]
        }
        testID="tab2Txt"
      >
        Owner
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="rightTab"
  >
    <View
      style={
        Array [
          Object {
            "borderBottomRightRadius": 40,
            "borderTopRightRadius": 40,
          },
          Object {
            "alignItems": "center",
            "backgroundColor": "#E2E2E2",
            "borderColor": "#fff",
            "borderWidth": 1,
            "elevation": 3,
            "paddingHorizontal": 14,
            "paddingVertical": 8,
            "shadowColor": "#181725",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.22,
            "shadowRadius": 2.22,
          },
          false,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontFamily": "Poppins-Regular",
              "fontSize": 16,
            },
            false,
          ]
        }
        testID="tab3Txt"
      >
        Admin
      </Text>
    </View>
  </View>
</View>
`;
